yarn-project/p2p/src/services/reqresp/agg_reqresp.ts

```diff
@@ -0,0 +1,659 @@
+      for (let batchStart = 0; batchStart < requests.length; batchStart += this.txsPerBatch) {
+        const batchEnd = Math.min(batchStart + this.txsPerBatch, requests.length);
+        const currentBatchSize = batchEnd - batchStart;
+
+        this.logger.debug(`Processing batch ${batchStart}-${batchEnd - 1} (${currentBatchSize} requests)`);
+
+        // Track which requests in this batch still need responses
+        const pendingRequestIndices = new Set(Array.from({ length: currentBatchSize }, (_, i) => batchStart + i));
+
+        // Create batch sampler for the current batch
+        const batchSampler = new AggressiveBatchConnectionSampler(this.connectionSampler, maxPeers);
+
+        if (batchSampler.activePeerCount() === 0) {
+          this.logger.debug('No active peers to send requests to');
+          continue;
+        }
+
+        // Get all peers we'll be requesting from
+        const peers = batchSampler.getAllPeers();
+        this.logger.debug(`Sending batch to ${peers.length} peers`);
+
+        // Send all requests to all peers simultaneously
+        const peerPromises = peers.map(async peer => {
+          try {
+            const peerResults: { index: number; response: InstanceType<SubProtocolMap[SubProtocol]['response']> }[] =
+              [];
+
+            // Process each request in the batch sequentially for this peer
+            for (const requestIndex of pendingRequestIndices) {
+              // Skip this request if it has already been fulfilled by another peer
+              if (!pendingRequestIndices.has(requestIndex)) {
+                continue;
+              }
+
+              const response = await this.sendRequestToPeer(peer, subProtocol, requestBuffers[requestIndex]);
+
+              if (response.status !== ReqRespStatus.SUCCESS) {
+                this.logger.debug(
+                  `Request ${requestIndex} to peer ${peer.toString()} failed with status ${prettyPrintReqRespStatus(
+                    response.status,
+                  )}`,
+                );
+                continue;
+              }
+
+              if (response && response.data.length > 0) {
+                const object = subProtocolMap[subProtocol].response.fromBuffer(response.data);
+                const isValid = await responseValidator(requests[requestIndex], object, peer);
+
+                if (isValid) {
+                  peerResults.push({ index: requestIndex, response: object });
+
+                  // Mark this request as fulfilled
+                  pendingRequestIndices.delete(requestIndex);
+                }
+              }
+            }
+
+            return { peer, results: peerResults };
+          } catch (error) {
+            this.logger.debug(`Failed batch request to peer ${peer.toString()}:`, error);
+            batchSampler.removePeerAndReplace(peer);
+            return { peer, results: [] };
+          }
+        });
+
+        const allResults = await Promise.allSettled(peerPromises);
```

Comment:
This will wait for all promises to resolve, so I guess it might wait for up to `timeoutMs = 10000` in case of slow/bad peer